ESPS/general/src/cepanal/cepanal.1
.\" Copyright (c) 1997 Entropic Research Laboratory, Inc. All rights reserved.
.\" @(#)cepanal.1	1.1 29 Apr 1997 ERL
.ds ]W (c) 1993 Entropic Research Laboratory, Inc.
.if t .ds pi \(*p
.if n .ds pi pi
.TH  CEPANAL 1\-ESPS 29 Apr 1997
.SH NAME

cepanal \- cepstrum and phase of sampled data
.SH SYNOPSIS
.B
cepanal
[
.BI \-l " frame_len"
] [
.BI \-o " order"
] [
.BI \-r " range"
] [
.BI \-w " window_type"
] [
.BI \-x " debug_level"
] [
.BI \-S " step"
]
.I input.sd
.I phase_out.fea
.I cepst_out.fea
.SH DESCRIPTION
.PP
This is one of a pair of programs that lets one
analyze a file of sampled data to obtain cepstrum and phase information,
modify the cepstrum,
and recombine the modified cepstrum with the original phase
to obtain a new sampled-data file.
This program analyzes the sampled data into cepstrum and phase.
The program
.IR cepsyn (1\-ESPS)
performs the inverse transformation, combining cepstrum and phase to
produce sampled data.
The input file,
.I input.sd,
must be a single-channel, real sampled-data file (FEA_SD or equivalent).
.PP
.I cepanal
forms the input into equally spaced frames of fixed length (see
.BR "\-l" ", " "\-S"
in the O\s-2PTIONS\s+2 section below),
optionally windows the data (see option
.BR "\-w" ),
performs a Fast Fourier Transform on each frame,
and computes the complex logarithm of the Fourier transform.
It saves the phase (the imaginary part of the logarithm)
in the output file
.I phase_out.fea.
It takes the inverse Fourier transform of the log magnitude
(the real part of the complex logarithm) to obtain the cepstrum
and outputs the result in
.I cepst_out.fea.
The output cepstrum is essentially the same as the output of
.sp
.nf
.RS
fftcep \-R \-F input.sd output.fea
.RE
.fi
.sp
(see
.IR fftcep (1\-ESPS))
with redundant values (nearly half the total number)
dropped from the end of each record.
.PP
The input frames have length
.I frame_len
(see option
.BR "\-l" ).
The initial point of the first frame is determined by the
\fB\-r\fP option.
Initial points of any subsequent frames follow at equal intervals
.I step
(see option
.BR \-S ).
Thus the 3 cases
.IR step " < " frame_len,
.IR step " = " frame_len,
and
.IR step " > " frame_len
correspond to overlapping frames, exactly abutted frames,
and frames separated by gaps.
.PP
The number of frames is the minimum sufficient to cover the specified
range (see option
.BR "\-r" ).
The last frame in the file may overrun the range,
in which case a warning is printed.
If a frame overruns the end of the file, it is filled out with zeros.
.PP
The output phase is not unwrapped, is between \-\*(pi and \*(pi,
and may have jumps with sizes of nearly 2\*(pi.
These discontinuities don't matter when the phase is used as input to
.IR cepsyn (1\-ESPS).
.PP
The output files are ordinary ESPS FEA files, not of any special FEA
subtype, and each contains a single field with data type FLOAT.
The field in
.I phase_out.fea
is named
.I phase;
its length is
.IR "n" "/2 + 1",
where
.I "n"
is the transform length (see option
.BR \-o ).
The phase values correspond to frequencies
that run in equal steps from 0 to
.IR "sf" "/2",
where
.I "sf"
is the sampling frequency.
The field in
.I cepst_out.fea
is named
.I cepstrum;
its length is also one greater than half the transform length
.I n
(see
.BR "\-o" ).
The cepstrum values correspond to quefrencies
that run in equal steps from 0 to
.RI "(" "n" "/2)(1/" "sf" ")."
.PP
If
.I input.sd
is "\-", standard input is read.
Either
.I phase_out.fea
or
.I cepst_out.fea,
but not both, may be "\-", for standard output.
.SH OPTIONS
.PP
The following options are supported.  Default values are shown in brackets.
.TP
.BI \-l " frame_len" "\fR [(transform length)]"
The length of each frame.
This is also the length of the data window applied to each frame (see option
.BI "\-w" ).
If the option is omitted or 0, a default value is used:
the transform length determined by the FFT order (see option
.BR "\-o" ).
.TP
.BI \-o " order" "\fR [10]\fP"
The order of the FFT used in the computations;
the transform length is
.RI 2^ order
(2 to the
.IR order -th
power).
This is also the default length of the frames
into which the input sampled-data stream is divided (see option
.BR "\-l" ).
If the frame length is less than the transform length,
each frame is padded with zeros.
(This is done after windowing\*-see option
.BI "\-w" ).
If the frame length is greater than the transform length,
each frame is truncated after windowing.
The default order, 10, corresponds to a transform length of 1024
and output fields of length 513.
.TP
.BI \-r " first" : "last" "\fR [1:(last in file)]"
.TP
.BI \-r " first" :+ "incr"
.TP
.BI \-r " first"
In the first form, a pair of unsigned integers specifies the range of
sampled data to analyze.
If
.I first
is omitted, the default value of 1 is used.  If
.I last
is omitted, the range extends to the end of the file.
If 
.IR last " = " first " + " incr,
the second form (with the plus sign) specifies the same range as the
first form.
The third form (omitting the colon) specifies a single sample.
If the range doesn't end on a frame boundary, it is extended to make
up a full last frame.
If a frame overruns the end of the file, it is filled out with zeros.
.TP
.BI \-w " window_type" "\fR [RECT]"
The name of the data window to apply to the data in each frame.
If the option is omitted, no windowing is done
(equivalent to a rectangular window with amplitude one).
Possible window types include
rectangular ("RECT"), Hamming ("HAMMING"), Hanning ("HANNING"),
cos^4 ("COS4"), and triangular ("TRIANG");
see the
.IR window (3-ESPSsp)
manual page for the complete list.
.TP
.BI \-x " debug_level \fR[0]\fP"
A positive value specifies that debugging output is to be printed on
the standard error output.
Larger values result in more output.
For the default value, 0, there is no output.
.TP
.BI \-S " step" "\fR [(frame length)]"
Initial points of consecutive frames differ by this number of samples.
If the option is omitted or 0, a default equal to
.I frame_len
is used (resulting in exactly abutted frames).
.SH ESPS PARAMETERS
.PP
This program does not access a parameter file.
.SH ESPS COMMON
.PP
The ESPS common file is not read.
.PP
If Common processing is enabled,
and the output file is not standard output,
the program writes the Common parameters
.I prog,
.I filename,
.I start,
and
.I nan
to record the program's name, the name of the input file,
the starting sample in the input file,
and the number of samples in the specified range.
(A value of 0 for
.I nan
implies that no explicit end of range was specified,
so that the program continued processing
to the end of the input file.)
.PP
ESPS Common processing may be disabled
by setting the environment variable USE_ESPS_COMMON to
.I off.
The default ESPS Common file is
.I espscom
in the user's home directory.
This may be overridden
by setting the environment variable ESPSCOM to the desired path.
User feedback of Common processing is determined
by the environment variable ESPS_VERBOSE, with 0 causing no feedback
and increasing levels causing increasingly detailed feedback.
If ESPS_VERBOSE is not defined, a default value of 3 is assumed.
.SH ESPS HEADERS
.PP
The program reads the header item
.I record_freq
from the input header, and it uses
.IR update_waves_gen (3\-ESPS)
to process the items
.I start_time
and
.I record_freq
in the standard way.
.PP
In addition to
.I start_time
and
.I record_freq,
the program add header items
.I start,
.I nan,
.I sf,
.I fft_length,
.I frmlen,
.I step,
and
.I window_type
to both output file headers.
The values of
.I start
and
.I nan
give the starting sample in the input file
and the number of samples in the specified range.
(A value of 0 for
.I nan
implies that no explicit end of range was specified,
so that the program continued processing
to the end of the input file.)
The value of
.I sf
(``sampling frequency'') records the value of
.I record_freq
from the input file.
The value of
.I fft_length
is the transform length, computed from the order.
The values of
.I frmlen
and
.I step
are the frame length and step (see the options
.B \-l
and
.BR \-S ).
The item
.I window_type
is a coded item that records the window type (see
.BR \-w ).
.PP
The programs defines fields
.I phase
in the header of
.I phase_out.fea
and
.I cepstrum
in the header of
.I cepst_out.fea.
Like most ESPS programs, it includes copies of the input file header
in the ouput headers,
and it saves a copy of the command line as a comment in each output
header.
.SH FUTURE CHANGES
.PP
.SH EXAMPLES
.PP
.SH ERRORS AND DIAGNOSTICS
.PP
If an unknown option is specified,
or if the number of file names is wrong,
.I cepanal
prints a synopsis of command-line usage and exits.
.PP
The program exits with an error message if any of the following are true:
the input file name is the same as one of the output file names
(but not "\-");
the output file names are the same;
the input file contains multichannel or complex sampled data;
a negative FFT order is specified;
a starting sample less than 1 is specified;
an empty range of records is specified;
no samples can be read from the input file.
.PP
The program prints a message and continues
if any of the following conditions are encountered:
the indicated input range does not end on a frame boundary;
the end of the input file is reached when an incomplete frame
has been read;
the end of the input file is reached prematurely
according to an explicitly indicated end of range.
.SH BUGS
.PP
None known.
.SH REFERENCES
.PP
.SH "SEE ALSO"
.PP
.na
.IR cepsyn (1\-ESPS),
.IR fft (1\-ESPS),
.IR fftcep (1\-ESPS),
.br
.IR fftinv (1\-ESPS)
.IR update_waves_gen (3\-ESPS)
.ad
.SH AUTHOR
.PP
Rod Johnson


ESPS/general/src/cepanal/cepanal.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1997 Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  Rod Johnson
 * Checked by:
 * Revised by:
 *
 * Brief description:  real cepstrum and phase factors from sample data
 */

static char *sccs_id = "@(#)cepanal.c	1.1	29 Apr 1997	ERL";

#define VERSION	"1.1"
#define DATE	"29 Apr 1997"
#define PROG	"cepanal"

int    debug_level = 0;

/* INCLUDE FILES */

#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <esps/esps.h>
#include <esps/fea.h>
#include <esps/feasd.h>
#include <esps/window.h>
#include <esps/func.h>
#include <esps/op.h>
#include <esps/types.h>
#include <esps/range_switch.h>
#include <esps/array.h>
#include <esps/get_fft.h>

/* LOCAL CONSTANTS */

#define DEF_ORDER   10			/* default order of FFT */

/* LOCAL MACROS */

#define SYNTAX \
USAGE(PROG " [-l frame_len] [-o order] [-r range] [-w window_type]\n" \
      "\t[-x debug_level] [-S step] input.sd phase_out.fea cepst_out.fea")

#define ERROR(text) { \
  (void) fprintf(stderr, "%s: %s - exiting\n", PROG, (text)); \
  exit(1);}

#define REQUIRE(test, text) {if (!(test)) ERROR(text)}

#define STRCMP(a, op, b) (strcmp((a), (b)) op 0)

/* SYSTEM FUNCTIONS AND VARIABLES */

extern int	getopt();			/* parse command options */
extern int	optind;				/* used by getopt */
extern char	*optarg;			/* used by getopt */

/* ESPS FUNCTIONS AND VARIABLES */

extern char	*get_cmd_line(int argc, char **argv);

/* LOCAL TYPEDEFS AND STRUCTURES */

/* LOCAL FUNCTION DECLARATIONS */

static void	pr_farray(char *text, long n, float *arr);
static void	pr_fcarray(char *text, long n, float_cplx *arr);

/* STATIC (LOCAL) GLOBAL VARIABLES */
 
/* MAIN PROGRAM */

int
main(int   argc,
     char  **argv)
{
    int		    ch;			/* command-line option letter */

    long	    i;			/* loop index */

    char	    *o_arg = NULL;	/* -o option argument */
    int		    order;		/* order of FFT */
    long	    translen;		/* FFT length */

    char	    *l_arg = NULL;	/* -l option argument */
    long	    framelen;		/* frame length */
    long	    maxlen;		/* translen or framelen, whichever is
					 * greater (length of temp arrays) */

    char	    *S_arg = NULL;	/* -S option argument */
    long	    step;		/* increment between frame endpoints */

    char	    *r_arg = NULL;	/* -r option argument */ 
    long	    startrec;		/* first sample to process */
    long	    endrec;		/* last sample to process */
    long	    nrec;		/* number of samples to process */
    long	    nan;		/* like nrec, but 0 means range extends
					 * from startrec to the end of file */

    char	    *w_arg = NULL;	/* -w option argument */
    int		    win;		/* window type code */

    long	    nfrm;		/* number of frames to process */
    long	    ifrm;		/* index of current frame */

    char	    *sdname;		/* input file name (sampled data) */
    FILE	    *sdfile;		/* input file */
    struct header   *sdhd;		/* input file header */
    struct feasd    *sdrec;		/* input data record structure */
    float	    *sddata;		/* input data array */
    int		    nchan;		/* number of input data channels */
    double	    sf;			/* sampling frequency */

    char	    *phname;		/* first (phase) output file name */
    FILE	    *phfile;		/* phase output file */
    struct header   *phhd;		/* phase file header */
    long	    phlen;		/* length of phase output field */
    struct fea_data *phrec;		/* phase output data record */
    float	    *phdata;		/* phase output data array */

    char	    *cepname;		/* second output file name (cepst) */
    FILE	    *cepfile;		/* cepstrum output file */
    struct header   *cephd;		/* cepstrum file header */
    long	    ceplen;		/* length of cepstrum output field */
    struct fea_data *ceprec;		/* cepstrum output data record */
    float	    *cepdata;		/* cepstrum output data array */

    long	    nvalid;		/* number of valid samples in
					 * buffer (returned by get_feasd
					 * functions) */
    long	    minvalid;		/* value returned by get_feasd_orecs
					 * when called at end of file */
    int		    eof;		/* end of input file encountered? */
    float	    *x, *y;		/* temp storage for FFT and inverse
					 * FFT computation (real and
					 * imaginary parts) */
    float_cplx	    *z;			/* FFT result (complex form) */
    float_cplx	    *logz;		/* Complex log of FFT */

    /*
     * PARSE COMMAND-LINE OPTIONS.
     */

    while ((ch = getopt(argc, argv, "l:o:r:w:x:S:")) != EOF)
        switch (ch)
	{
	case 'l':
	    l_arg = optarg;
	    break;
	case 'o':
	    o_arg = optarg;
	    break;
	case 'r':
	    r_arg = optarg;
	    break;
	case 'w':
	    w_arg = optarg;
	    break;
	case 'x':
	    debug_level = atoi(optarg);
	    break;
	case 'S':
	    S_arg = optarg;
	    break;
	default:
	    SYNTAX;
	}

    /*
     * PROCESS FILE NAMES.
     */

    if (argc - optind > 3)
    {
	fprintf(stderr, "%s: too many file names.\n", PROG);
	SYNTAX;
    }
    if (argc - optind < 3)
    {
	fprintf(stderr, "%s: not enough file names.\n", PROG);
	SYNTAX;
    }

    sdname = argv[optind++];
    phname = argv[optind++];
    cepname = argv[optind++];

    REQUIRE(STRCMP(sdname, ==, "-")
	    || STRCMP(sdname, !=, phname) && STRCMP(sdname, !=, cepname),
	    "Output file cannot be same as input file");

    REQUIRE(STRCMP(phname, !=, cepname),
	    "Output files cannot be the same");

    /*
     * OPEN AND CHECK INPUT FILE.
     */

    sdname = eopen(PROG, sdname, "r", FT_FEA, FEA_SD, &sdhd, &sdfile);

    if (debug_level >= 1)
	fprintf(stderr, "%s: input file = \"%s\".\n", PROG, sdname);

    nchan = (int) get_fea_siz("samples", sdhd, NULL, NULL);

    REQUIRE(nchan == 1,
	    "Multichannel data not supported");
    REQUIRE(!is_field_complex(sdhd, "samples"),
	    "Complex input data not supported");

    sf = *get_genhd_d("record_freq", sdhd);

    /*
     * PROCESS OPTIONS.
     */

    /* ORDER, FRAMELEN, STEP */

    order =
	(o_arg != NULL) ? atoi(o_arg) : DEF_ORDER;

    REQUIRE(order >= 0,
	    "Order of FFT can't be negative");

    translen = LROUND(pow(2.0, (double) order));

    framelen =
	(l_arg != NULL) ? atol(l_arg) : translen;
    if (framelen == 0)
	framelen = translen;

    step =
	(S_arg != NULL) ? atol(S_arg) : framelen;
    if (step == 0)
	step = framelen;

    if (debug_level >= 1)
	fprintf(stderr,
		"%s: order = %d, translen = %ld, "
		"framelen = %ld, step = %ld.\n",
		PROG, order, translen, framelen, step);

    /* RANGE */

    startrec = 1;
    endrec = LONG_MAX;
    if (r_arg != NULL)
	lrange_switch(r_arg, &startrec, &endrec, NO);

    REQUIRE(startrec >= 1,
	    "Can't start before beginning of file");
    if (endrec == LONG_MAX)
    {
	nrec = LONG_MAX;
	nfrm = LONG_MAX;
    }
    else
    {
	REQUIRE(endrec >= startrec,
		"Empty range of records specified");

	nrec = endrec - startrec + 1; /* number of samples in range */

	/* Find the number of frames corresponding to "nrec" samples.
	 * For the first frame, "framelen" samples are read.
	 * For each additional frame, if any, "step" additional samples
	 * are read.
	 * Partial frames are rounded _up_ to the nearest whole frame.
	 */
	nfrm =
	    (nrec <= framelen) ? 1
		: (step <= framelen) ? 2 + (nrec - framelen - 1) / step
		    : 1 + (nrec - 1) / step;

	if (framelen + (nfrm - 1) * step > nrec)
	    fprintf(stderr,
		    "%s: Last frame will overrun range by %ld points.\n",
		    PROG, framelen + (nfrm - 1) * step - nrec);
    }

    nan = (nrec == LONG_MAX) ? 0 : nrec;

    if (debug_level >= 1)
	fprintf(stderr,
		"%s: startrec = %ld, endrec = %ld,\n"
		"\tnrec = %ld, nan = %ld, nfrm = %ld.\n",
		PROG, startrec, endrec, nrec, nan, nfrm);

    /* WINDOW */

    win =
	(w_arg != NULL) ? win_type_from_name(w_arg) : WT_RECT;
    REQUIRE(win != WT_NONE, "Invalid window type");
    REQUIRE(win != WT_KAISER, "Sorry, KAISER window not supported");
    REQUIRE(win != WT_ARB, "Sorry, ARB window not supported");

    if (debug_level >= 1)
	fprintf(stderr, "%s: win = %s.\n", PROG, window_types[win]);

    /*
     * SET UP OUTPUT FILES
     */

    /* PHASE FILE */

    phname = eopen(PROG, phname, "w", NONE, NONE, NULL, &phfile);

    if (debug_level >= 1)
	fprintf(stderr,
		"%s: output file for phase = \"%s\".\n", PROG, phname);

    phhd = new_header(FT_FEA);
    add_source_file(phhd, sdname, sdhd);
    (void) strcpy(phhd->common.prog, PROG);
    (void) strcpy(phhd->common.vers, VERSION);
    (void) strcpy(phhd->common.progdate, DATE);
    add_comment(phhd, get_cmd_line(argc, argv));
    update_waves_gen(sdhd, phhd,
		     (float) (startrec + framelen/2), (float) step);
    /* framelen/2 is integer division
     * (for compatibility with fft & fftcep) */

    *add_genhd_l("start", NULL, 1, phhd) = startrec;
    *add_genhd_l("nan", NULL, 1, phhd) = nan;
    *add_genhd_d("sf", NULL, 1, phhd) = sf;
    *add_genhd_l("fft_length", NULL, 1, phhd) = translen;
    *add_genhd_l("frmlen", NULL, 1, phhd) = framelen;
    *add_genhd_l("step", NULL, 1, phhd) = step;
    *add_genhd_e("window_type", NULL, 1, window_types, phhd) = win;
    phlen = translen/2 + 1;
    (void) add_fea_fld("phase", phlen, 1, NULL, FLOAT, NULL, phhd);

    if (debug_level >= 1)
	fprintf(stderr, "%s: size of phase field is %ld.\n", PROG, phlen);
    if (debug_level >= 1)
	fprintf(stderr, "%s: writing phase file header.\n", PROG);

    write_header(phhd, phfile);
    
    /* CEPSTRUM FILE */

    cepname = eopen(PROG, cepname, "w", NONE, NONE, NULL, &cepfile);

    if (debug_level >= 1)
	fprintf(stderr,
		"%s: output file for cepstrum = \"%s\".\n", PROG, cepname);

    cephd = new_header(FT_FEA);
    add_source_file(cephd, sdname, sdhd);
    (void) strcpy(cephd->common.prog, PROG);
    (void) strcpy(cephd->common.vers, VERSION);
    (void) strcpy(cephd->common.progdate, DATE);
    add_comment(cephd, get_cmd_line(argc, argv));
    update_waves_gen(sdhd, cephd,
		     (float) (startrec + framelen/2), (float) step);
    /* framelen/2 is integer division
     * (for compatibility with fft & fftcep) */

    *add_genhd_l("start", NULL, 1, cephd) = startrec;
    *add_genhd_l("nan", NULL, 1, cephd) = nan;
    *add_genhd_d("sf", NULL, 1, cephd) = sf;
    *add_genhd_l("fft_length", NULL, 1, cephd) = translen;
    *add_genhd_l("frmlen", NULL, 1, cephd) = framelen;
    *add_genhd_l("step", NULL, 1, cephd) = step;
    *add_genhd_e("window_type", NULL, 1, window_types, cephd) = win;
    ceplen = translen/2 + 1;
    (void) add_fea_fld("cepstrum", ceplen, 1, NULL, FLOAT, NULL, cephd);

    if (debug_level >= 1)
	fprintf(stderr, "%s: size of cepstrum field is %ld.\n", PROG, ceplen);
    if (debug_level >= 1)
	fprintf(stderr, "%s: writing cepstrum file header.\n", PROG);

    write_header(cephd, cepfile);

    /*
     * ALLOCATE RECORD STRUCTURES AND STORAGE
     */

    if (debug_level >= 1)
	fprintf(stderr, "%s: allocating storage.\n", PROG);

    sdrec = allo_feasd_recs(sdhd, FLOAT, framelen, NULL, NO);
    sddata = (float *) sdrec->data;

    maxlen = MAX(framelen, translen);
    x = (float *) arr_alloc(1, &maxlen, FLOAT, 0);
    y = (float *) arr_alloc(1, &maxlen, FLOAT, 0);
    z = NULL;			/* first call of arr_op will allocate */
    logz = NULL;		/* first call of arr_func will allocate */

    phrec = allo_fea_rec(phhd);
    phdata = (float *) get_fea_ptr(phrec, "phase", phhd);

    ceprec = allo_fea_rec(cephd);
    cepdata = (float *) get_fea_ptr(ceprec, "cepstrum", cephd);

    /*
     * MAIN LOOP
     */

    /* Set up for eof test.  In the case of overlapping frames
     * (framelen > step), the function get_feasd_orecs will return
     * framelen - step when it fails to read any new samples, since
     * it obtains that many valid samples by shifting samples already
     * in the buffer.
     */
    minvalid = (framelen <= step) ? 0 : framelen - step;
    eof = NO;

    /* Get first frame. */

    fea_skiprec(sdfile, startrec - 1, sdhd);
    nvalid = get_feasd_recs(sdrec, 0, framelen, sdhd, sdfile);
    REQUIRE(nvalid > 0,
	    "No samples read from input");
    ifrm = 1;

    if (debug_level >= 1)
	fprintf(stderr, "%s: beginning of main loop.\n", PROG);

    /* Loop, processing a frame, writing the results, and reading a new
     * frame, until we reach either the end of the specified range or the
     * end of the file.
     */

    do
    {
	if (debug_level >= 2)
	    fprintf(stderr, "\n%s: FRAME %ld\n", PROG, ifrm);

	if (nvalid < framelen)
	{
	    eof = YES;
	    fprintf(stderr,
		    "%s: EOF in frame %ld; only %ld points read.\n",
		    PROG, ifrm, nvalid);
	}

	if (debug_level >= 2)
	    pr_farray("input sampled data",
		      (debug_level == 2) ? MIN(framelen, 10) : framelen,
		      sddata);

	/* WINDOW INPUT DATA */

	window(framelen, sddata, x, win, NULL); /* when win is WT_RECT,
						 * this just copies */
	for (i = 0; i < framelen; i++)
	{
	    y[i] = 0.0;		/* supply imag part */
	}
	for (i = framelen; i < translen; i++)
	{
	    x[i] = 0.0;
	    y[i] = 0.0;
	}

	/* GET FFT OF WINDOWED DATA */

	if (debug_level >= 2)
	    pr_farray("windowed sampled data",
		      (debug_level == 2) ? MIN(framelen, 10) : framelen,
		      sddata);

	get_rfft(x, y, order);

	if (debug_level >= 2)
	{
	    pr_farray("FFT of input (real part)",
		      (debug_level == 2) ? MIN(translen, 10) : translen,
		      x);
	    pr_farray("FFT of input (imag part)",
		      (debug_level == 2) ? MIN(translen, 10) : translen,
		      y);
	}

	/* TAKE COMPLEX LOG OF FFT */

	z = (float_cplx *) arr_op(OP_CPLX, translen,
				  (char *) x, FLOAT, (char *) y, FLOAT,
				  (char *) z, FLOAT_CPLX);
	logz = (float_cplx *) arr_func(FN_LOG, translen,
				       (char *) z, FLOAT_CPLX,
				       (char *) logz, FLOAT_CPLX);

	if (debug_level >= 2)
	    pr_fcarray("complex log of FFT",
		       (debug_level == 2) ? MIN(framelen, 10) : framelen,
		       logz);

	/* GET PHASE AND WRITE IT OUT */

	(void) arr_func(FN_IM, phlen,
			(char *) logz, FLOAT_CPLX, (char *) phdata, FLOAT);
	put_fea_rec(phrec, phhd, phfile);

	if (debug_level >= 2)
	    pr_farray("phase",
		      (debug_level == 2) ? MIN(phlen, 10) : phlen,
		      phdata);

	/* GET CEPSTRUM AND WRITE IT OUT */

	(void) arr_func(FN_RE, translen,
			(char *) logz, FLOAT_CPLX, (char *) x, FLOAT);
	for (i = 0; i < translen; i++)
	    y[i] = 0.0;

	if (debug_level >= 2)
	    pr_farray("log of magnitude of FFT",
		      (debug_level == 2) ? MIN(translen, 10) : translen,
		      x);

	get_fft_inv(x, y, order);

	for (i = 0; i < ceplen; i++)
	    cepdata[i] = x[i];
	put_fea_rec(ceprec, cephd, cepfile);

	if (debug_level >= 2)
	{
	    pr_farray("cepstrum (Re)",
		      (debug_level == 2) ? MIN(ceplen, 10) : ceplen,
		      x);
	    pr_farray("cepstrum (Im)",
		      (debug_level == 2) ? MIN(ceplen, 10) : ceplen,
		      y);
	}

	/* We go back to process a new frame as long as
	 * (1) we haven't yet processed the number of frames required for
	 *     the specified range,
	 * (2) we didn't hit the end of file while reading the current
	 *     frame, and
	 * (3) we get at least one new sample when we try to read the new
	 *     frame (i.e. nvalid > minvalid; see comment above on
	 *     initialization of minvalid).
	 */
    } while (++ifrm <= nfrm
	     && !eof
	     && ((nvalid = get_feasd_orecs(sdrec, 0,
					   framelen, step, sdhd, sdfile))
		 > minvalid));

    if (debug_level >= 2)
	fprintf(stderr, "\n");
    if (debug_level >= 1)
	fprintf(stderr, "%s: end of main loop; ifrm = %ld.\n",
		PROG, ifrm);

    if (nfrm != LONG_MAX && ifrm <= nfrm)
	fprintf(stderr, "%s: Premature EOF; only %ld frames read.\n",
		PROG, ifrm - 1);

    /*
     * WRITE COMMON AND EXIT
     */

    if (phfile != stdout && cepfile != stdout)
    {
	REQUIRE(startrec <= INT_MAX
		&& putsym_i("start", (int) startrec) != -1,
		"Can't write \"start\" to Common.");
	REQUIRE(nan <= INT_MAX
		&& putsym_i("nan", (int) nan) != -1,
		"Can't write \"nan\" to Common.");
	REQUIRE(putsym_s("prog", PROG) != -1,
		"Can't write \"prog\" to Common.");
	REQUIRE(putsym_s("filename", sdname) != -1,
		"Can't write \"filename\" to Common.");
    }

    exit(0);
    return 0;
}


/* LOCAL FUNCTION DEFINITIONS */

/*
 * For debug printout of float arrays
 */

static void
pr_farray(char    *text,
	  long    n,
	  float   *arr)
{
    long    i;

    fprintf(stderr, "%s: %s -- %d points:\n", PROG, text, n);
    for (i = 0; i < n; i++)
    {
	fprintf(stderr, "%g ", arr[i]);
	if (i%5 == 4 || i == n - 1)
	    fprintf(stderr, "\n");
    }
}

/*
 * For debug printout of float_cplx arrays
 */

static void
pr_fcarray(char		*text,
	   long		n,
	   float_cplx	*arr)
{
    long    i;

    fprintf(stderr, "%s: %s -- %d points:\n", PROG, text, n);
    for (i = 0; i < n; i++)
    {
	fprintf(stderr, "[%g %g] ", arr[i].real, arr[i].imag);
	if (i%2 == 1 || i == n - 1)
	    fprintf(stderr, "\n");
    }
}

